flocker/control/test/test_persistence.py

```diff
@@ -249,3 +333,202 @@ def test_complex_keys(self):
+    external_port=ST_PORTS
+)
+LINKS = st.builds(
+    Link,
+    local_port=ST_PORTS,
+    remote_port=ST_PORTS,
+    alias=SIMPLE_TEXT
+)
+APPLICATIONS = st.builds(
+    Application, name=SIMPLE_TEXT, image=IMAGES,
+    ports=st.sets(PORTS, max_size=10),
+    links=st.sets(LINKS, max_size=10),
+    environment=st.dictionaries(keys=SIMPLE_TEXT, values=SIMPLE_TEXT),
+    memory_limit=NONE_OR_INT,
+    cpu_shares=NONE_OR_INT,
+    running=st.booleans()
+)
+FILEPATHS = st.text(alphabet=string.printable).map(FilePath)
+VOLUMES = st.builds(
+    AttachedVolume, manifestation=MANIFESTATIONS, mountpoint=FILEPATHS)
+APPLICATIONS_WITH_VOLUMES = st.tuples(
+    APPLICATIONS, VOLUMES).map(lambda (a, v): a.set(volume=v))
+
+
+def _build_node(applications):
+    # All the manifestations in `applications`.
+    app_manifestations = set(app.volume.manifestation for app in applications)
+    # A set that contains all of those, plus an arbitrary set of
+    # manifestations.
+    dataset_ids = frozenset(
+        app.volume.manifestation.dataset_id
+        for app in applications if app.volume
+    )
+    manifestations = (
+        st.sets(MANIFESTATIONS.filter(
+            lambda m: m.dataset_id not in dataset_ids))
+        .map(pset)
+        .map(lambda ms: ms.union(app_manifestations))
+        .map(lambda ms: dict((m.dataset.dataset_id, m) for m in ms)))
+    return st.builds(
+        Node, uuid=UUIDS,
+        applications=st.just(applications),
+        manifestations=manifestations)
+
+
+NODES = st.sets(APPLICATIONS_WITH_VOLUMES).map(
+    lambda apps: pset(dict((
+        app.volume.manifestation.dataset_id, app) for app in apps).values())
+).flatmap(_build_node)
+
+
+DEPLOYMENTS = st.builds(
+    Deployment, nodes=st.sets(NODES, min_size=0, max_size=3)
+)
+
+
+class ConfigurationMigrationTests(SynchronousTestCase):
+    """
+    Tests for ``ConfigurationMigration`` class that performs individual
+    configuration upgrades.
+
+    There should be
+    """
+    @given(DEPLOYMENTS)
+    def test_upgrade_configuration_v1_latest(self, deployment):
+        """
+        Test a range of generated configurations (deployments) can be
+        upgraded from v1 to the latest configuration version.
+
+        This test will need updating for each new configuration version
+        introduced to reflect the expected configuration format
+        (``expected_configuration``) after a successful upgrade.
+        """
+        source_json = wire_encode(deployment)
+        upgraded_json = migrate_configuration(
+            1, _CONFIG_VERSION, source_json, ConfigurationMigration)
+        upgraded_config = wire_decode(upgraded_json)
+        expected_configuration = Configuration(
+            version=_CONFIG_VERSION, deployment=deployment
+        )
+        self.assertEqual(upgraded_config, expected_configuration)
+
+    @given(st.tuples(
+        st.integers(min_value=1, max_value=2),
+        st.integers(min_value=2, max_value=2)
+    ))
```

Comment:
Since `versions` is never used as a tuple, this is better specified as `@given(integers(...), integers(...))`, and then writing the test as `test_upgrade_configuration_versions(self, from_version, to_version)`.